ql/src/java/org/apache/hadoop/hive/ql/parse/CalcitePlanner.java

```diff
@@ -2349,6 +2362,50 @@ private RelNode applyMaterializedViewRewriting(RelOptPlanner planner, RelNode ba
       return basePlan;
     }
 
+    private RelOptMaterialization copyMaterializationToNewCluster(RelOptCluster optCluster, RelOptMaterialization materialization) {
+      final RelNode viewScan = materialization.tableRel;
+      final RelNode newViewScan = HiveMaterializedViewUtils.copyNodeNewCluster(
+              optCluster, viewScan);
+      return new RelOptMaterialization(newViewScan, materialization.queryRel, null,
+              materialization.qualifiedTableName);
+    }
+
+    private boolean isMaterializedViewRewritingByTextEnabled() {
+      return conf.getBoolVar(ConfVars.HIVE_MATERIALIZED_VIEW_ENABLE_AUTO_REWRITING_QUERY_TEXT) &&
+              mvRebuildMode == MaterializationRebuildMode.NONE &&
+              !getQB().isMaterializedView() && !ctx.isLoadingMaterializedView() && !getQB().isCTAS() &&
+              getQB().getIsQuery() &&
+              getQB().hasTableDefined();
+    }
+
+    private RelNode applyMaterializedViewRewritingByText(RelNode calciteGenPlan, RelOptCluster optCluster) {
+      unparseTranslator.applyTranslations(ctx.getTokenRewriteStream(), EXPANDED_QUERY_TOKEN_REWRITE_PROGRAM);
+      String expandedQueryText = ctx.getTokenRewriteStream()
+              .toString(EXPANDED_QUERY_TOKEN_REWRITE_PROGRAM, ast.getTokenStartIndex(), ast.getTokenStopIndex());
+      try {
+        List<RelOptMaterialization> relOptMaterializationList = db.getMaterialization(
+                expandedQueryText, getTablesUsed(calciteGenPlan), getTxnMgr());
+        for (RelOptMaterialization relOptMaterialization : relOptMaterializationList) {
+          try {
+            Table hiveTableMD = extractTable(relOptMaterialization);
+            if (db.validateMaterializedViewsFromRegistry(
```

Comment:
Another validation is needed here to confirm whether current user has access to such MV. See `HiveMaterializedViewUtils.checkPrivilegeForMaterializedViews` (https://github.com/apache/hive/commit/cf4463e2d049ceaf931d21e2ae8182f50d56b43d)

Please, add a test such as the one included in the https://github.com/apache/hive/commit/cf4463e2d049ceaf931d21e2ae8182f50d56b43d .

Reply:
added check, and tests